core/src/net/address.cpp

```diff
@@ -0,0 +1,190 @@
+#ifdef _WIN32
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#pragma comment(lib, "Ws2_32.lib")
+#else
+#include <arpa/inet.h>
+#include <netdb.h>
+#endif
+
+#include <algorithm>
+#include <cstring>
+
+#include <cubos/core/net/address.hpp>
+#include <cubos/core/reflection/external/cstring.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+#include <cubos/core/reflection/traits/string_conversion.hpp>
+#include <cubos/core/reflection/type.hpp>
+
+using namespace cubos::core::net;
+using cubos::core::memory::Opt;
+
+const Address Address::Any((uint8_t)0, 0, 0, 0);
+const Address Address::LocalHost((uint8_t)127, 0, 0, 1);
+
+CUBOS_REFLECT_IMPL(cubos::core::net::Address)
+{
+    using cubos::core::reflection::StringConversionTrait;
+    using cubos::core::reflection::Type;
+
+    return Type::create("cubos::core::net::Address")
+        .with(StringConversionTrait{
+            [](const void* instance) { return static_cast<const Address*>(instance)->toString(); },
+            [](void* instance, const std::string& string) {
+                auto addr = Address::from(string);
+                if (addr.contains())
+                {
+                    *static_cast<Address*>(instance) = addr.value();
+                    return true;
+                }
+                return false;
+            }});
+}
+
+/// @brief Inits needed resources.
+static void init()
+{
+#ifdef _WIN32
+    WSADATA wsa;
+    CUBOS_ASSERT(WSAStartup(MAKEWORD(2, 2), &wsa) == 0, " WSAStartup failed: {}", WSAGetLastError());
+#endif
+}
+
+/// @brief Cleans up resources.
+static void cleanup()
+{
+#ifdef _WIN32
+    WSACleanup();
+#endif
+}
+
+Address::Address()
+    : mType(IPType::IPv4)
+{
+    init();
+}
+
+Address::Address(uint8_t a, uint8_t b, uint8_t c, uint8_t d)
+    : mAddr{.v4Addr = static_cast<uint32_t>(a) << 24 | static_cast<uint32_t>(b) << 16 | static_cast<uint32_t>(c) << 8 |
+                      static_cast<uint32_t>(d)}
+    , mType(IPType::IPv4)
+{
+    init();
+}
+
+Address::Address(uint32_t address)
+    : mAddr{.v4Addr = address}
+    , mType(IPType::IPv4)
+{
+    init();
```

Comment:
Maybe its better to call `init` only when you actually use the functions, and clean up immediately after, as you do in `from`.